sda/websocket.go

```diff
@@ -0,0 +1,366 @@
+package sda
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"time"
+
+	"reflect"
+	"strings"
+
+	"sync"
+
+	"github.com/dedis/cothority/log"
+	"github.com/dedis/cothority/network"
+	"github.com/dedis/protobuf"
+	"github.com/gorilla/websocket"
+	"gopkg.in/tylerb/graceful.v1"
+)
+
+// WebSocket handles incoming client-requests using the WebSocket
+// protocol for JavaScript-compatibility. When making a new WebSocket,
+// it will listen one port above the ServerIdentity-port-#.
+type WebSocket struct {
+	services  map[string]Service
+	server    *graceful.Server
+	mux       *http.ServeMux
+	startstop chan bool
+	started   bool
+	sync.Mutex
+}
+
+const (
+	// WebSocketErrorPathNotFound indicates the path has not been registered
+	WebSocketErrorPathNotFound = 4000
+	// WebSocketErrorProtobufDecode indicates an error in decoding the protobuf-packet
+	WebSocketErrorProtobufDecode = 4001
+	// WebSocketErrorProtobufEncode indicates an error in encoding the return packet
+	WebSocketErrorProtobufEncode = 4002
+	// WebSocketErrorInvalidErrorCode indicates the service returned
+	// an invalid error-code
+	WebSocketErrorInvalidErrorCode = 4003
+)
+
+// NewWebSocket opens a webservice-listener one port above the given
+// ServerIdentity.
+func NewWebSocket(si *network.ServerIdentity) *WebSocket {
+	w := &WebSocket{
+		services:  make(map[string]Service),
+		startstop: make(chan bool),
+	}
+	webHost, err := getWebHost(si)
+	log.ErrFatal(err)
+	w.mux = http.NewServeMux()
+	w.server = &graceful.Server{
+		Timeout: 100 * time.Millisecond,
+		Server: &http.Server{
+			Addr:    webHost,
+			Handler: w.mux,
+		},
+	}
+	return w
+}
+
+// Start listening on the port.
+func (w *WebSocket) Start() {
+	w.Lock()
+	w.started = true
+	w.Unlock()
+	log.Lvl3("Starting to listen on", w.server.Server.Addr)
+	go func() {
+		w.server.ListenAndServe()
+		w.startstop <- false
+	}()
+	w.startstop <- true
+}
+
+// RegisterService saves the service as being able to handle messages.
+func (w *WebSocket) RegisterService(service string, s Service) error {
+	w.services[service] = s
+	return nil
+}
+
+// RegisterMessageHandler for a service. Requests from a client to
+// "ws://service/path" will be forwarded to the corresponding
+// ProcessClientRequest of the 'service'.
+func (w *WebSocket) RegisterMessageHandler(service, path string) error {
```

Comment:
Chronologically, in the `newCoSiService`-method, the `RegisterMessageHandler` is called, then the `RegisterService` once the `newCoSiService` is done.

When a message comes in, the first part of the path is used to look up the service and find the corresponding `ProcessClientRequest`, and the second part of the path indicates what message-handler needs to be used.

But I think I can get rid of the `RegisterService`-method.